### PR TITLE
test: add coverage for all edge cases of the TracerBuilder

### PR DESCRIPTION
**Pull request by eugypalu**
Resolves: #1082 

# Pull Request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [x] Testing

# What is the new behavior?

Added tests to cover all failure cases of the TracerBuilder using a mocked EthereumProvider.

# Does this introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -147,6 +147,7 @@ dojo-test-utils = { git = 'https://github.com/dojoengine/dojo', tag = "v1.0.0-al
 katana-primitives = { git = 'https://github.com/dojoengine/dojo', tag = "v1.0.0-alpha.2", default-features = false, features = [
   "serde",
 ], optional = true }
+mockall = { version = "0.13.0", default-features = false }
```
[the remark this answers is not in the document]
Comment (eugypalu) [reply Cargo.toml]: I agree, fixed. thanks

File: src/tracing/builder.rs [hunk 2]
```diff
@@ -220,3 +220,83 @@ impl<P: EthereumProvider + Send + Sync + Clone> TracerBuilder<P, Pinned> {
         env
     }
 }
+
+#[cfg(test)]
```
[the remark this answers is not in the document]
Comment (eugypalu) [reply src/tracing/builder.rs:227]: Updated, added comments to the tests.

**Review by tcoratger — APPROVED** [review]

**Comment by tcoratger** [comment]
pending @greged93 :)

File: src/tracing/builder.rs [hunk 3]
```diff
@@ -220,3 +220,119 @@ impl<P: EthereumProvider + Send + Sync + Clone> TracerBuilder<P, Pinned> {
...
+        mock_provider
+            .expect_transaction_by_hash()
+            .returning(|_| Ok(Some(Transaction { block_number: None, ..Default::default() })));
+
+        // Create a TracerBuilder with the mock provider
+        let builder = TracerBuilder::new(Arc::new(&mock_provider)).await.unwrap();
+        // Attempt to use the builder with a specific transaction hash, expecting an error
+        let result = builder.with_transaction_hash(B256::repeat_byte(0)).await;
+        // Check that the result is an UnknownBlock error
+        assert!(matches!(result, Err(EthApiError::UnknownBlock(_))));
```
Comment (greged93) [inline src/tracing/builder.rs:324]: hmm we should update the error, this doesn't look right. Can you update it in the tracer builder?

File: src/tracing/builder.rs [hunk 4]
```diff
@@ -220,3 +220,119 @@ impl<P: EthereumProvider + Send + Sync + Clone> TracerBuilder<P, Pinned> {
...
+        // Create a mock Ethereum provider
+        let mut mock_provider = MockEthereumProviderStruct::new();
+        // Expect the chain_id call to return an error
+        mock_provider.expect_chain_id().returning(|| Err(TransactionError::InvalidChainId.into()));
+        // Attempt to create a TracerBuilder with the mock provider, expecting an error
+        let result = TracerBuilder::new(Arc::new(&mock_provider)).await;
+        // Check that the result is an error
+        assert!(result.is_err());
+    }
+}
```
Comment (greged93) [inline src/tracing/builder.rs:338]: I think we are missing a test for this
https://github.com/kkrt-labs/kakarot-rpc/blob/6052a52fc2c575327850595221df92ab368587c4/src/tracing/builder.rs#L186

Comment (greged93) [inline src/tracing/builder.rs:338]: I think we are also missing a test for this
https://github.com/kkrt-labs/kakarot-rpc/blob/6052a52fc2c575327850595221df92ab368587c4/src/tracing/builder.rs#L167

Comment (eugypalu) [reply src/tracing/builder.rs:347]: added, thanks

File: src/tracing/builder.rs [hunk 5]
```diff
@@ -220,3 +220,119 @@ impl<P: EthereumProvider + Send + Sync + Clone> TracerBuilder<P, Pinned> {
...
+        // Create a TracerBuilder with the mock provider
+        let builder = TracerBuilder::new(Arc::new(&mock_provider)).await.unwrap();
+        // Attempt to use the builder with a specific block hash, expecting an error
+        let result = builder.with_block_id(BlockId::Hash(B256::repeat_byte(1).into())).await;
+        // Check that the result is an UnknownBlock error
+        assert!(matches!(result, Err(EthApiError::UnknownBlock(_))));
+    }
+
+    #[tokio::test]
+    async fn test_tracer_builder_with_transaction_hash_failure() {
```
[the remark this answers is not in the document]
Comment (eugypalu) [reply src/tracing/builder.rs]: I agree, removed

File: src/tracing/builder.rs [hunk 6]
```diff
@@ -220,3 +220,119 @@ impl<P: EthereumProvider + Send + Sync + Clone> TracerBuilder<P, Pinned> {
...
+// such as unknown blocks, not found transactions, and invalid chain IDs.
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::test_utils::mock_provider::MockEthereumProviderStruct;
+    use reth_primitives::U64;
+    use reth_rpc_types::Transaction;
+    use std::sync::Arc;
+    #[tokio::test]
+    async fn test_tracer_builder_with_block_id_failure_block_id() {
```
[the remark this answers is not in the document]
Comment (eugypalu) [reply src/tracing/builder.rs:235]: For this test and the next one, I am not sure if they should be removed because they allow testing the error of the `block` function, which would otherwise remain uncovered. What do you think?

[the remark this answers is not in the document]
Comment (greged93) [reply src/tracing/builder.rs:235]: ok I got confused because of the test itself, I think you should be able to have this work with the below modification which translates more what you want to test

File: src/tracing/builder.rs [hunk 7]
```diff
@@ -220,3 +220,128 @@ impl<P: EthereumProvider + Send + Sync + Clone> TracerBuilder<P, Pinned> {
...
+        // Create a mock Ethereum provider
+        let mut mock_provider = MockEthereumProviderStruct::new();
+        // Expect the chain_id call to return 1
+        mock_provider.expect_chain_id().returning(|| Ok(Some(U64::from(1))));
+        // Expect the block_by_number call to return an error for an unknown block
+        mock_provider.expect_block_by_number().returning(|_, _| {
+            Err(EthApiError::UnknownBlock(BlockHashOrNumber::Number(u64::try_from(U64::from(1)).unwrap())))
+        });
```
Comment (greged93) [inline src/tracing/builder.rs:243]:
```suggestion
        mock_provider.expect_block_by_number().returning(|_, _| {
            Ok(None)
        });
```

File: src/tracing/builder.rs [hunk 8]
```diff
@@ -220,3 +220,119 @@ impl<P: EthereumProvider + Send + Sync + Clone> TracerBuilder<P, Pinned> {
...
+        // Create a TracerBuilder with the mock provider
+        let builder = TracerBuilder::new(Arc::new(&mock_provider)).await.unwrap();
+        // Attempt to use the builder with a specific block ID, expecting an error
+        let result = builder.with_block_id(BlockId::Number(1.into())).await;
+        // Check that the result is an UnknownBlock error
+        assert!(matches!(result, Err(EthApiError::UnknownBlock(_))));
+    }
+
+    #[tokio::test]
+    async fn test_tracer_builder_with_block_id_failure_transaction_hash() {
```
[the remark this answers is not in the document]
Comment (greged93) [reply src/tracing/builder.rs:254]: Should we rename this? it doesn't test transaction hash failure right?

File: src/tracing/builder.rs [hunk 9]
```diff
@@ -220,3 +220,128 @@ impl<P: EthereumProvider + Send + Sync + Clone> TracerBuilder<P, Pinned> {
...
+        // Create a mock Ethereum provider
+        let mut mock_provider = MockEthereumProviderStruct::new();
+        // Expect the chain_id call to return 1
+        mock_provider.expect_chain_id().returning(|| Ok(Some(U64::from(1))));
+        // Expect the block_by_hash call to return an error for an unknown block
+        mock_provider
+            .expect_block_by_hash()
+            .returning(|_, _| Err(EthApiError::UnknownBlock(B256::repeat_byte(0).into())));
```
Comment (greged93) [inline src/tracing/builder.rs:262]:
```suggestion
        mock_provider
            .expect_block_by_hash()
            .returning(|_, _| Ok(None));
```

File: src/tracing/builder.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (eugypalu) [reply src/tracing/builder.rs]: Renamed to `test_tracer_builder_block_failure_with_none_block_hash`, and tested the `block` function directly

File: src/tracing/builder.rs [hunk 10]
```diff
@@ -220,3 +220,124 @@ impl<P: EthereumProvider + Send + Sync + Clone> TracerBuilder<P, Pinned> {
...
+        // Create a TracerBuilder with the mock provider
+        let builder = TracerBuilder::new(Arc::new(&mock_provider)).await.unwrap();
+        // Attempt to use the builder with a specific transaction hash, expecting an error
+        let result = builder.with_transaction_hash(B256::repeat_byte(0)).await;
+        // Check that the result is an UnknownBlock error
+        assert!(matches!(result, Err(EthApiError::TransactionNotFound(_))));
+    }
+
+    #[tokio::test]
+    async fn test_tracer_builder_failure() {
```
Comment (greged93) [inline src/tracing/builder.rs:305]: I think this can go, it doesn't really test the builder behaviour, just that the provider throws. Let's just fix this and ok to merge for me!

Comment (eugypalu) [reply src/tracing/builder.rs]: removed, thanks

**Review by greged93 — DISMISSED** [review]

**Review by tcoratger — APPROVED** [review]
